### PR TITLE
Makefile: refactor test recipe

### PR DESCRIPTION
**Pull request by straight-shoota**
The `spec` recipe builds `spec/all_spec.cr` which combines `std_spec`, `compiler_spec` and `primitives_spec` into a single spec suite. This is huge and requires a lot of resources, both for building as well as running the specs. It can easily exhaust many build system's memory capabilities.

I don't think we're using it *anywhere* whatsoever. And it's really not recommended due to the intense resource requirements.

But since it's there and mentioned as the top test recipe, users (especially package maintainers) are likely to pick it up.

This patch includes the following changes:

* Rename `spec` to `all_spec` and add a note that it's expensive.
* Add `test` as a replacement which runs the parts (`std_spec`, `compiler_spec` and `primitives_spec`) individually.
* Update the example make commands

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -80,9 +84,8 @@ check_llvm_config = $(eval \
 .PHONY: all
 all: crystal ## Build all files (currently crystal only) [default]
 
-.PHONY: spec
-spec: $(O)/all_spec ## Run all specs
-	$(O)/all_spec $(SPEC_FLAGS)
+.PHONY: test ## Run tests
```
Comment (HertzDevil) [inline Makefile:87]: For `make help` the `##` should go on the same line as the target itself, not the phony one (alternatively `test:` can be declared on an additional line like what `smoke_test` does)

**Review by HertzDevil — COMMENTED** [review]
`Makefile.win` needs the same changes

**Comment by asterite** [comment]
Why the name `test`? It's a bit confusing given that tests in Crystal are called specs.

By the way, I used to use `make spec` all the time to make sure all specs pass before I submit a PR, but I think it's okay to rename this to `all_spec`. But maybe `test` should be `spec`?

**Comment by straight-shoota** [comment]
@asterite The idea for renaming is to make clear that it does something different than `make spec`. The purpose is still the same of course, and maybe that's good enough, but the mode is different.

In any case, `test` is a more general term and by convention a typical standard target for Makefiles.

**Comment by asterite** [comment]
> The idea for renaming is to make clear that it does something different than make spec

Oh, then I'm confused. What does `make spec` do now with this PR?

**Comment by straight-shoota** [comment]
The `spec` target doesn't exist anymore.
I was expecting `make spec` to fail, but it doesn't. The `spec` folder exists and make recognizes that. It just reports there's nothing to do.

Then it's probably better to define `spec` as a phony target. It could either do what `test` does (implicitly change the meaning of `make spec` to run specs individually) or we could let it fail with an error message to use either `make test` or `make all_spec` instead.

**Comment by asterite** [comment]
> Add `test` as a replacement which runs the parts (std_spec, compiler_spec and primitives_spec) individually.

I don't understand why we can't change `spec` to do that instead. Why rename `spec` to `test`?

**Comment by straight-shoota** [comment]
Like I said in https://github.com/crystal-lang/crystal/pull/12979#issuecomment-1397206221 the idea was to make `make spec` fail so you have to decide whether you really want the `all_spec` behaviour with a single executable (then you'd need to make it explicit). If not, change it to `test` instead.

**Comment by asterite** [comment]
What would be the dangers of naming it `spec`? Let's consider a few users:
1. Me, or others wanting to check all specs pass before pushing code: it works like before, except that it consumes less resources, and it might be even faster because we don't need to wait for all specs to be compiled (say: the first batch fails, we get feedback earlier)
2. Package maintainers: they probably check the error code to see if all specs pass. Nothing changes for them.
3. Others? I'm not sure.

With this change, if some packages relied on `spec` you are now forcing everyone to update their code or flow... for no real reason.

**Comment by straight-shoota** [comment]
Okay, let's keep the name `spec`.

I'll still keep `test` as an alias because that's the makefile convention.

**Review by HertzDevil — APPROVED** [review]

**Review by HertzDevil — APPROVED** [review]

**Pull request merged**